From 93b6d20ca1638c4e65b5b67a98f43ebefbe004ed Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Sat, 16 Jul 2016 18:11:51 +0100
Subject: [PATCH 5/5] wm: Trial inclusion of force unredirect configuration

On certain devices, disabling unredirect of windows can result in performance
improvements in full screen mode, i.e. when gaming. However, in certain
configurations this can actually result in a negative user experience, i.e.
through excessive tearing. As such, we expose this as an option, and not a
default, for those looking for the best performance whilst gaming.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 wm/com.solus-project.budgie.wm.gschema.xml |  6 ++++++
 wm/wm.vala                                 | 22 ++++++++++++++++++++++
 2 files changed, 28 insertions(+)

diff --git a/wm/com.solus-project.budgie.wm.gschema.xml b/wm/com.solus-project.budgie.wm.gschema.xml
index e93b6fd..2658c01 100644
--- a/wm/com.solus-project.budgie.wm.gschema.xml
+++ b/wm/com.solus-project.budgie.wm.gschema.xml
@@ -14,6 +14,12 @@
       <description>This key overrides the key in org.gnome.mutter when running Budgie.</description>
     </key>
 
+    <key type="b" name="force-unredirect">
+      <default>false</default>
+      <summary>Forcibly disable screen unredirection</summary>
+      <description>Disables the screen unredirection which may improve performance with certain drivers.</description>
+    </key>
+
     <key type="as" name="toggle-raven">
       <default><![CDATA[['<Super>A']]]></default>
       <summary>The binding to use to toggle Raven applets view</summary>
diff --git a/wm/wm.vala b/wm/wm.vala
index 4922369..8e850bb 100644
--- a/wm/wm.vala
+++ b/wm/wm.vala
@@ -15,6 +15,7 @@ namespace Budgie {
 public static const string MUTTER_EDGE_TILING  = "edge-tiling";
 public static const string MUTTER_MODAL_ATTACH = "attach-modal-dialogs";
 public static const string MUTTER_BUTTON_LAYOUT = "button-layout";
+public static const string WM_FORCE_UNREDIRECT = "force-unredirect";
 public static const string WM_SCHEMA           = "com.solus-project.budgie-wm";
 
 public static const bool CLUTTER_EVENT_PROPAGATE = false;
@@ -105,6 +106,8 @@ public class BudgieWM : Meta.Plugin
     WindowMenu? winmenu = null;
     LoginDRemote? logind_proxy = null;
 
+    private Settings? wm_settings = null;
+
     HashTable<Meta.WindowActor?,AnimationState?> state_map;
 
     static construct
@@ -327,6 +330,10 @@ public class BudgieWM : Meta.Plugin
 
         var display = screen.get_display();
 
+        this.settings = new Settings(WM_SCHEMA);
+        this.settings.connect("changed", this.on_wm_schema_changed);
+        this.on_wm_schema_changed(WM_FORCE_UNREDIRECT);
+
         state_map = new HashTable<Meta.WindowActor?,AnimationState?>(GLib.direct_hash, GLib.direct_equal);
 
         Meta.Prefs.override_preference_schema(MUTTER_EDGE_TILING, WM_SCHEMA);
@@ -395,6 +402,21 @@ public class BudgieWM : Meta.Plugin
         keyboard.hook_extra();
     }
 
+    private void on_wm_schema_changed(string key)
+    {
+        if (key != WM_FORCE_UNREDIRECT) {
+            return;
+        }
+        bool enab = this.settings.get_boolean(key);
+
+        var screen = this.get_screen();
+        if (enab) {
+            Meta.Util.disable_unredirect_for_screen(screen);
+        } else {
+            Meta.Util.enable_unredirect_for_screen(screen);
+        }
+    }
+
     public override void show_window_menu(Meta.Window window, Meta.WindowMenuType type, int x, int y)
     {
         if (type != Meta.WindowMenuType.WM) {
-- 
2.9.1

